yt/yql/agent/yql_agent.cpp

```diff
@@ -183,11 +212,18 @@ class TYqlAgent
 
     void OnDynamicConfigChanged(
         const TYqlAgentDynamicConfigPtr& /*oldConfig*/,
-        const TYqlAgentDynamicConfigPtr& /*newConfig*/) override
-    { }
+        const TYqlAgentDynamicConfigPtr& newConfig) override
+    {
+        DynamicConfig_ = newConfig;
+    }
 
     TFuture<std::pair<TRspStartQuery, std::vector<TSharedRef>>> StartQuery(TQueryId queryId, const TString& user, const TReqStartQuery& request) override
     {
+        if (ActiveQueriesGuard::Get() >= DynamicConfig_->MaxSimultaneousQueries) {
+            YT_LOG_INFO("Query was throttled (QueryId: %v, User: %v)", queryId, user);
+            THROW_ERROR_EXCEPTION(NYqlClient::EErrorCode::Throttled, "Query was throttled");
+        }
+
         YT_LOG_INFO("Starting query (QueryId: %v, User: %v)", queryId, user);
 
         return BIND(&TYqlAgent::DoStartQuery, MakeStrong(this), queryId, user, request)
```

Comment:
Next line is `.AsyncVia(ThreadPool_->GetInvoker())`

As long as `MaxSimultaneousQueries`(128) is lower than `YqlThreadCount`(256), the delay between checking `ActiveQueriesGuard::Get() >= DynamicConfig_->MaxSimultaneousQueries` and increasing `ActiveQueries` should be minimal (`ThreadPool_` queue should be empty)

In the opposite case this throttling will stop working as we will never reach values above `YqlThreadCount`. So we may want to add this check (`YqlThreadCount >= MaxSimultaneousQueries`) somewhere

But even with this small delay between checking ActiveQueries (line 222) and increasing it (line 287) we may see a situation with a long queue. This can be illustrated with this unlikely example:
1. 1000 queries is simultaneously dispatched to yql agent and start executing `StartQuery`
2. All of them execute line 222 and continue forward to line 229
3. 256 of them start executing in the `ThreadPool_`, the rest are in ThreadPool's queue
4. 256 queries increase ActiveQueries counter, finally preventing new queries to be added to `ThreadPool_`

This is usually resolved by using conditional increase, as shown in [this stackoverflow question](https://stackoverflow.com/questions/13778110/is-there-atomic-increment-with-the-check-preconditions-that-the-atomic-value-wa)

So I suggest using it to atomically check that `ActiveQueries < MaxSimultaneousQueries` and increase `ActiveQueries` if this is true:
1. We add a check that `YqlThreadCount` is strictly more than `MaxSimultaneousQueries`, to make sure that we have at least 1 free thread in `ThreadPool_`
2. In DoStartQuery in guard constructor we do:
```
IsTaken_ = true;
auto activeQueries = ActiveQueries_.load();
do
{
  if (activeQueries >= MaxSimultaneousQueries) {
    IsTaken_ = false;
    return;
  }
} while (!ActiveQueries_.compare_exchange_weak(activeQueries, activeQueries + 1));
```
3. In DoStartQuery after guard creation we do `if (!guard.IsTaken_) THROW_THROTTLED_ERROR`
4. In guard destructor we do nothing if `!IsTaken_`

This can still lead to queue in `ThreadPool_`. But it should be quickly emptied by at least 1 thread that will be taking requrests from there and throwing throttled exceptions.

One solution to avoid it is to conditionally increase ActiveQueries in `StartQuery` and decrease it after `DoStartQuery` finish. But it will add more complexity and will require passing a guard from `StartQuery` to `DoStartQuery`, so I suggest sticking to a solution described above as a balanced one